File: src/main/java/org/tindalos/principle/infrastructure/reporters/SAPViolationsReporter.java
package org.tindalos.principle.infrastructure.reporters;

import java.util.List;

import org.tindalos.principle.domain.core.Package;
import org.tindalos.principle.domain.coredetector.ViolationsReporter;
import org.tindalos.principle.domain.detector.sap.SAPResult;

public class SAPViolationsReporter implements ViolationsReporter<SAPResult> {

	public String report(SAPResult result) {
		List<Package> outlierPackages = result.getOutlierPackages();
		String sectionLine = "==============================================================";
		StringBuffer sb = new StringBuffer("\n" + sectionLine + "\n");
		sb.append("\tStable Abstractions Principle violations ("+outlierPackages.size()+" of allowed "+result.getThreshold()+" )\t");
		sb.append("\n" + sectionLine + "\n");

		if (outlierPackages.isEmpty()) {
			sb.append("No violations.\n");
		} else {
			for (Package outlierPackage : outlierPackages) {
				sb.append(print(outlierPackage) + "\n");
			}
		}
		sb.append(sectionLine + "\n");
		return sb.toString();
	}

	private String print(Package outlierPackage) {
		return outlierPackage + " " + outlierPackage.distance();
	}

	public Class<SAPResult> getType() {
		return SAPResult.class;
	}

}

File: src/main/java/org/tindalos/principle/infrastructure/reporters/SAPViolationsReporter.scala
package org.tindalos.principle.infrastructure.reporters

import org.tindalos.principle.domain.core.Package
import org.tindalos.principle.domain.detector.sap.SAPResult
import org.tindalos.principle.domain.resultprocessing.reporter.AnalysisResultsReporter

object SAPViolationsReporter {

  def report(result: SAPResult):AnalysisResultsReporter.Report = {
    val outlierPackages = result.outlierPackages
    val sectionLine = "=============================================================="
    val sb = new StringBuffer("\n" + sectionLine + "\n")
    sb.append("\tStable Abstractions Principle violations (" + outlierPackages.length + " of allowed " + result.threshold + " )\t")
    sb.append("\n" + sectionLine + "\n")

    if (outlierPackages.isEmpty) {
      sb.append("No violations.\n")
    } else {
      outlierPackages foreach { outlierPackage =>
        sb.append(print(outlierPackage) + "\n")
      }
    }
    sb.append(sectionLine + "\n")
    sb.toString()
  }

  private def print(outlierPackage: Package) = outlierPackage + " " + outlierPackage.distance

}